File: prototype/rails/src/test/domain/shop/checkout_form_test.rb
require "test_helper"

class CheckoutFormTest < ActiveSupport::TestCase
  CheckoutForm = Domain::Shop::CheckoutForm

  test "a filled form is complete" do
    form = CheckoutForm.from_input(email: " Ada@Example.Test ", shipping: shipping)

    assert form.complete?
    assert_empty form.missing_parts
    assert_equal "Ada@Example.Test", form.email
    assert_equal "London", form.shipping.city
  end

  test "the second address line is optional" do
    form = CheckoutForm.from_input(email: "ada@example.test", shipping: shipping(line2: ""))

    assert form.complete?
    assert_nil form.shipping.line2
  end

  test "a blank shipping part is missing" do
    form = CheckoutForm.from_input(email: "ada@example.test", shipping: shipping(city: "   ", country: nil))

    refute form.complete?
    assert_equal %i[city country], form.missing_parts
  end

  test "an address that is not an email is incomplete" do
    refute CheckoutForm.from_input(email: "ada", shipping: shipping).complete?
  end

  test "an absent shipping part is missing" do
    form = CheckoutForm.from_input(email: "ada@example.test", shipping: {})

    refute form.complete?
    assert_equal CheckoutForm::REQUIRED_PARTS, form.missing_parts
  end

  private

  def shipping(**overrides)
    {
      name: "Ada Lovelace", line1: "12 Analytical Way", line2: "Flat 3",
      city: "London", region: "Greater London", postal_code: "EC1A 1BB", country: "GB"
    }.merge(overrides)
  end
end

File: prototype/node/src/app/core/shop/checkout-form.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { parseCheckoutForm } from './checkout-form.ts'

function shipping(overrides: Record<string, string | null> = {}): Record<string, string | null> {
  return {
    name: 'Ada Lovelace',
    line1: '12 Analytical Way',
    line2: 'Flat 3',
    city: 'London',
    region: 'Greater London',
    postalCode: 'EC1A 1BB',
    country: 'GB',
    ...overrides,
  }
}

test('a filled form parses into a complete value', () => {
  const parsed = parseCheckoutForm({ email: ' Ada@Example.Test ', shipping: shipping() })

  assert.equal(parsed.ok, true)
  if (!parsed.ok) return
  assert.equal(parsed.value.email, 'Ada@Example.Test')
  assert.equal(parsed.value.shipping.city, 'London')
})

test('the second address line is optional', () => {
  const parsed = parseCheckoutForm({ email: 'ada@example.test', shipping: shipping({ line2: '' }) })

  assert.equal(parsed.ok, true)
  if (!parsed.ok) return
  assert.equal(parsed.value.shipping.line2, null)
})

test('a blank shipping part is an error beside that field, not a value', () => {
  const parsed = parseCheckoutForm({
    email: 'ada@example.test',
    shipping: shipping({ city: '   ', country: null }),
  })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, { city: 'Enter the city.', country: 'Enter the country.' })
})

test('an address that is not an email is an error beside the email field', () => {
  const parsed = parseCheckoutForm({ email: 'ada', shipping: shipping() })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, { email: 'Enter a valid email address.' })
})

test('an absent shipping part is an error beside every required field', () => {
  const parsed = parseCheckoutForm({ email: 'ada@example.test', shipping: {} })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.deepEqual(parsed.errors, {
    name: 'Enter the full name.',
    line1: 'Enter the address.',
    city: 'Enter the city.',
    region: 'Enter the region.',
    postalCode: 'Enter the postal code.',
    country: 'Enter the country.',
  })
})

test('a refused form hands back what was entered, trimmed, for the page to show again', () => {
  const parsed = parseCheckoutForm({
    email: '  ada@example.test  ',
    shipping: shipping({ city: '  London  ', country: '' }),
  })

  assert.equal(parsed.ok, false)
  if (parsed.ok) return
  assert.equal(parsed.entered.email, 'ada@example.test')
  assert.equal(parsed.entered.shipping.city, 'London')
  assert.equal(parsed.entered.shipping.country, '')
})
